{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "cc1776ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This notebook uses the output of the rule\n",
      "--- 63_eval_read_align.smk::combine_read_depth_stats\n",
      "as input, and produces a supplementary figure\n",
      "summarizing long-read coverage in the de novo\n",
      "assemblies for autosomes, X and Y at different\n",
      "MAPQ thresholds.\n",
      "\n",
      "Repository directory:  /home/ebertp/work/code/marschall-lab/project-male-assembly\n",
      "Execution directory:  /home/ebertp/work/code/marschall-lab/project-male-assembly/notebooks/plotting/readcov\n",
      "Working directory:  /home/ebertp/work/projects/sig_chry/paper\n",
      "Output directory:  /home/ebertp/work/projects/sig_chry/paper/output/figures\n",
      "=================================\n"
     ]
    }
   ],
   "source": [
    "import datetime as dt\n",
    "import pathlib as pl\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpat\n",
    "\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "datetime_now = dt.datetime.now()\n",
    "ts = datetime_now.strftime(\"D%y%m%dT%H%M\")\n",
    "\n",
    "TS_PLOT = False\n",
    "TS_FILE = True\n",
    "\n",
    "desc=\"\"\"\n",
    "This notebook uses the output of the rule\n",
    "--- 63_eval_read_align.smk::combine_read_depth_stats\n",
    "as input, and produces a supplementary figure\n",
    "summarizing long-read coverage in the de novo\n",
    "assemblies for autosomes, X and Y at different\n",
    "MAPQ thresholds.\n",
    "\"\"\"\n",
    "\n",
    "create_plots = False\n",
    "save_plots = False\n",
    "\n",
    "print(desc)\n",
    "\n",
    "repo_dir = pl.Path('/home/ebertp/work/code/marschall-lab/project-male-assembly').resolve(strict=True)\n",
    "exec_dir = pl.Path('.').resolve(strict=True)\n",
    "wd_dir = pl.Path('/home/ebertp/work/projects/sig_chry/paper').resolve(strict=True)\n",
    "out_dir = pl.Path('/home/ebertp/work/projects/sig_chry/paper/output/figures').resolve(strict=True)\n",
    "\n",
    "print('Repository directory: ', repo_dir)\n",
    "print('Execution directory: ', exec_dir)\n",
    "print('Working directory: ', wd_dir)\n",
    "print('Output directory: ', out_dir)\n",
    "print('=================================')\n",
    "\n",
    "table = pd.read_csv(\n",
    "    wd_dir.joinpath(\n",
    "        \"stats/read_cov\",\n",
    "        \"ALL-SAMPLES.READS_aln-to_HIFIRW.ONTUL.na.wg.cov-stats.tsv\"\n",
    "    ),\n",
    "    header=0,\n",
    "    index_col=[\"location\", \"statistic\", \"reads\", \"min_mapq\", \"input_set\"],\n",
    "    sep=\"\\t\"\n",
    ")\n",
    "\n",
    "drop_samples = [\n",
    "    \"HG02666\",\n",
    "    \"NA18989\",\n",
    "    \"HG01457\",\n",
    "    \"NA19384\",\n",
    "    \"NA24385\"\n",
    "]\n",
    "\n",
    "table = table.loc[~table[\"sample\"].isin(drop_samples), :].copy()\n",
    "\n",
    "box_plots = []\n",
    "box_positions = []\n",
    "\n",
    "labels = []\n",
    "label_pos = []\n",
    "quantiles = []\n",
    "\n",
    "box_pos = 1\n",
    "\n",
    "if create_plots:\n",
    "    for reads in [\"HIFIRW\", \"ONTUL\"]:\n",
    "        for loc, input_set in [(\"auto\", \"wg\"), (\"chrX\", \"wg\"), (\"chrY\", \"wg\"), (\"chrY\", \"noYHET\")]:\n",
    "            if box_pos > 1:\n",
    "                box_pos += 2\n",
    "            for mapq in [0, 10]:\n",
    "                values = table.xs(\n",
    "                    (loc, input_set, mapq, reads, \"median_cov\"),\n",
    "                    level=[\"location\", \"input_set\", \"min_mapq\", \"reads\", \"statistic\"]\n",
    "                )\n",
    "                #labels.append(f\"{loc}_MQ{mapq}\")\n",
    "                box_plots.append(values[\"value\"])\n",
    "                box_positions.append(box_pos)\n",
    "                if mapq < 10:\n",
    "                    label_pos.append(box_pos + 0.5)\n",
    "                    if loc == \"auto\":\n",
    "                        label = \"Other\\n(not X or Y)\"\n",
    "                    elif loc == \"chrY\" and input_set == \"wg\":\n",
    "                        label = \"chrY\\n(full)\"\n",
    "                    elif loc == \"chrY\":\n",
    "                        label = \"chrY\\n(no HET)\"\n",
    "                    else:\n",
    "                        label = \"chrX\"\n",
    "                    labels.append(label)\n",
    "                box_pos += 1\n",
    "                quantiles.append([0.01,0.99])\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(10, 8))\n",
    "        violins = ax.violinplot(\n",
    "            box_plots,\n",
    "            positions=box_positions,\n",
    "            showmedians=True,\n",
    "            showextrema=False,\n",
    "            quantiles=quantiles\n",
    "        )\n",
    "        for idx, violin in enumerate(violins[\"bodies\"]):\n",
    "            if idx % 2 == 1:\n",
    "                violin.set_facecolor(\"cornflowerblue\")\n",
    "            else:\n",
    "                violin.set_facecolor(\"salmon\")\n",
    "        violins[\"cmedians\"].set_color(\"grey\")\n",
    "        violins[\"cquantiles\"].set_color(\"grey\")\n",
    "\n",
    "        ax.set_xticks(label_pos, labels, fontsize=12)\n",
    "        readset = \"HiFi\" if reads == \"HIFIRW\" else \"ONT-UL\"\n",
    "        ax.set_ylabel(f\"Median read depth per assembled contig\\n({readset} reads)\", fontsize=14)\n",
    "        ax.set_xlabel(\"Genomic location\", fontsize=14)\n",
    "        if reads == \"HIFIRW\":\n",
    "            ax.set_ylim(-1, 50)\n",
    "        else:\n",
    "            ax.set_ylim(-1, 70)\n",
    "\n",
    "        custom_legend = []\n",
    "        custom_legend.append(\n",
    "            mpat.Patch(facecolor=\"salmon\", label=\"MAPQ >= 0\", edgecolor=\"white\")\n",
    "        )\n",
    "        custom_legend.append(\n",
    "            mpat.Patch(facecolor=\"cornflowerblue\", label=\"MAPQ >= 10\", edgecolor=\"white\")\n",
    "        )\n",
    "        ax.legend(handles=custom_legend, fontsize=14)\n",
    "        ax.spines[\"top\"].set_visible(False)\n",
    "        ax.spines[\"right\"].set_visible(False)\n",
    "        if TS_PLOT:\n",
    "            ax.set_title(timestamp, pad=40, fontsize=10)\n",
    "\n",
    "        if save_plots:\n",
    "            out_base_name = f\"FigSX_{readset}-read-depth_auto-X-Y\"\n",
    "            if TS_FILE:\n",
    "                out_base_name += f\".{ts}\"\n",
    "            plt.savefig(\n",
    "                out_dir / pl.Path(out_base_name + '.pdf'),\n",
    "                bbox_inches='tight',\n",
    "                facecolor=\"w\", transparent=None,\n",
    "                #bbox_extra_artists=[cmap_legend]\n",
    "            )\n",
    "            plt.savefig(\n",
    "                out_dir / pl.Path(out_base_name + '.png'),\n",
    "                bbox_inches='tight',\n",
    "                facecolor=\"w\", transparent=None,\n",
    "                #bbox_extra_artists=[cmap_legend],\n",
    "                dpi=150\n",
    "            )\n",
    "\n",
    "        # reset everything\n",
    "        labels = []\n",
    "        label_pos = []\n",
    "        box_plots = []\n",
    "        box_pos = 1\n",
    "        box_positions = []\n",
    "        quantiles = []\n",
    "    \n",
    "# reduce table to a few data points\n",
    "# for inclusion in the supplement\n",
    "\n",
    "table.reset_index(drop=False, inplace=True)\n",
    "\n",
    "selector = (table[\"min_mapq\"] > 0) & (table[\"location\"] != \"wg\") & (table[\"statistic\"] == \"median_cov\")\n",
    "\n",
    "table = table.loc[selector, :].copy()\n",
    "\n",
    "table[\"loc_spec\"] = table[\"location\"] + \"_\" + table[\"input_set\"]\n",
    "\n",
    "joined_table = []\n",
    "\n",
    "for readset in [\"HIFIRW\", \"ONTUL\"]:\n",
    "    reduced = table.loc[table[\"reads\"] == readset, :].pivot_table(\n",
    "        index=\"sample\",\n",
    "        columns=\"loc_spec\",\n",
    "        values=\"value\",\n",
    "        aggfunc=\"median\"\n",
    "    )\n",
    "    reduced[\"reads\"] = \"HiFi\" if readset == \"HIFIRW\" else \"ONT-UL\"\n",
    "    joined_table.append(reduced)\n",
    "\n",
    "joined_table = pd.concat(joined_table, axis=0, ignore_index=False)\n",
    "joined_table.reset_index(drop=False, inplace=True)\n",
    "joined_table.sort_index(inplace=True)\n",
    "joined_table.sort_values([\"sample\", \"reads\"], inplace=True)\n",
    "\n",
    "joined_table = joined_table[\n",
    "    [\"sample\", \"reads\", \"auto_wg\", \"chrX_wg\", \"chrY_wg\", \"chrY_noYHET\"]\n",
    "]\n",
    "\n",
    "joined_table.columns = [\"sample\", \"reads\", \"autosomes\", \"chrX\", \"chrY\", \"chrY_noHET\"]\n",
    "\n",
    "table_out_name = \"TableSX_assm-read-cov.median.tsv\"\n",
    "with open(out_dir.parent.joinpath(\"tables\", table_out_name), \"w\") as table:\n",
    "    joined_table.to_csv(table, sep=\"\\t\", header=True, index=False)\n",
    "    table.write(f\"\\n## {ts}\\n## 63_eval_read_align.smk::combine_read_depth_stats\\n\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
