{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7c9dfbb9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This notebook uses the output of the rule\n",
      "--- 90_est_gaps.smk::merge_bubbles_node_infos\n",
      "as input, and produces supplementary figures\n",
      "illustrating PAR1 variation.\n",
      "\n",
      "Execution directory:  /home/ebertp/work/code/marschall-lab/project-male-assembly/notebooks/plotting/par1_graphs\n",
      "Working directory:  /home/ebertp/work/code/marschall-lab/project-male-assembly/notebooks/plotting/par1_graphs\n",
      "Output directory:  /home/ebertp/work/projects/sig_chry/paper/output/figures\n",
      "=================================\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pathlib as pl\n",
    "import collections as col\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# required to have text output as \"text fields\"\n",
    "# and not individual objects\n",
    "plt.rcParams['pdf.fonttype'] = 42\n",
    "plt.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "save_plots = True\n",
    "\n",
    "desc=\"\"\"\n",
    "This notebook uses the output of the rule\n",
    "--- 90_est_gaps.smk::merge_bubbles_node_infos\n",
    "as input, and produces supplementary figures\n",
    "illustrating PAR1 variation.\n",
    "\"\"\"\n",
    "\n",
    "print(desc)\n",
    "\n",
    "repo_dir = pl.Path('/home/ebertp/work/code/marschall-lab/project-male-assembly').resolve(strict=True)\n",
    "exec_dir = pl.Path('.').resolve(strict=True)\n",
    "wd_dir = exec_dir\n",
    "out_dir = pl.Path('/home/ebertp/work/projects/sig_chry/paper/output/figures').resolve(strict=True)\n",
    "\n",
    "print('Execution directory: ', exec_dir)\n",
    "print('Working directory: ', wd_dir)\n",
    "print('Output directory: ', out_dir)\n",
    "print('=================================')\n",
    "\n",
    "input_path = pl.Path(\"/home/ebertp/work/projects/sig_chry/paper/data/graphs\")\n",
    "\n",
    "color_lut = dict()\n",
    "\n",
    "y_par1 = 2458321\n",
    "y_xdr1 = 2726838\n",
    "y_xtr1 = 5914797\n",
    "\n",
    "BIN_SIZE = 10000\n",
    "BIN_LABEL = int(BIN_SIZE / 1e3)\n",
    "SHIFT_SIZE = 5000\n",
    "SHIFT_LABEL = int(SHIFT_SIZE / 1e3)\n",
    "\n",
    "for table in input_path.glob(\"*flat-bubbles.tsv\"):\n",
    "    table_spec = table.name.split(\".\")\n",
    "    backbone = table_spec[3]\n",
    "    num_samples = int(table_spec[1].strip(\"samples\"))\n",
    "    size_t = int(table_spec[4].strip(\"Lkbp\"))\n",
    "    if size_t != 1:\n",
    "        continue\n",
    "    if num_samples != 10:\n",
    "        continue\n",
    "    if backbone in [\"chrY\", \"chrYX\"]:\n",
    "        if backbone == \"chrY\":\n",
    "            backbone = \"Y\"\n",
    "            complement = None\n",
    "        else:\n",
    "            backbone = \"Y\"\n",
    "            complement = \"X\"\n",
    "    elif backbone in [\"chrX\", \"chrXY\"]:\n",
    "        if backbone == \"chrX\":\n",
    "            backbone = \"X\"\n",
    "            complement = None\n",
    "        else:\n",
    "            backbone = \"X\"\n",
    "            complement = \"Y\"\n",
    "    else:\n",
    "        raise ValueError(backbone)\n",
    "\n",
    "    if complement is None:\n",
    "        continue\n",
    "    y_par1_set = False\n",
    "    y_xdr1_set = False\n",
    "    y_xtr1_set = False\n",
    "\n",
    "    df = pd.read_csv(table, header=0, sep=\"\\t\")\n",
    "    df = df.loc[(df[\"is_source\"] == 0) & (df[\"is_sink\"] == 0), :]\n",
    "    last_bubble = df.loc[df[\"start\"] < 6e6, \"bubble_num\"].max()\n",
    "    last_end = max(df.loc[df[\"bubble_num\"] == last_bubble, \"end\"].max(), 6e6)\n",
    "    df = df.loc[df[\"end\"] < last_end + 1, :]\n",
    "    \n",
    "    pop_variation = []\n",
    "    \n",
    "    x_ticks = []\n",
    "    x_ticklabels = []\n",
    "\n",
    "    for win_idx, start in enumerate(np.arange(0, last_end + 1, step=SHIFT_SIZE)):\n",
    "        end = start + BIN_SIZE\n",
    "        if int(start) in [500000, 1000000, 1500000, 2000000, 3500000, 4000000, 4500000, 5000000, 5500000]:\n",
    "            x_ticks.append(win_idx)\n",
    "            label = str(round(start/1e6, 1)).strip(\"0\").rstrip(\".\")\n",
    "            x_ticklabels.append(label)\n",
    "        if backbone == \"X\":\n",
    "            if int(start) in [2500000, 3000000, 6000000]:\n",
    "                x_ticks.append(win_idx)\n",
    "                label = str(round(start/1e6, 1)).strip(\"0\").rstrip(\".\")\n",
    "                x_ticklabels.append(label)\n",
    "        if backbone == \"Y\":\n",
    "            if (start < y_par1 < end) and not y_par1_set:\n",
    "                x_ticks.append(win_idx)\n",
    "                x_ticklabels.append(\"PAR1\")\n",
    "                y_par1_set = True\n",
    "            if (start < y_xdr1 < end) and not y_xdr1_set:\n",
    "                x_ticks.append(win_idx)\n",
    "                x_ticklabels.append(\"XDR1\")\n",
    "                y_xdr1_set = True\n",
    "            if (start < y_xtr1 < end) and not y_xtr1_set:\n",
    "                x_ticks.append(win_idx)\n",
    "                x_ticklabels.append(\"XTR1\")\n",
    "                y_xtr1_set = True\n",
    "            \n",
    "        window = (start < df[\"end\"]) & (end > df[\"start\"])\n",
    "        win_variation = 0.\n",
    "        pop_var = col.Counter()\n",
    "        for bubble, nodes in df.loc[window, :].groupby(\"bubble_id\"):\n",
    "            total_allele_length = nodes[\"node_length\"].sum()\n",
    "            if total_allele_length < 1000:\n",
    "                continue\n",
    "            if (nodes[\"sample\"] == \"T2T.X\").all() or (nodes[\"sample\"] == \"T2T.Y\").all():\n",
    "                # skip over bubbles that are all ref\n",
    "                # probably something filtered after the\n",
    "                # initial construction of the minigraph?\n",
    "                continue\n",
    "            overlap = min(end, nodes[\"end\"].unique()) - max(start, nodes[\"start\"].unique())\n",
    "            if overlap == 0:\n",
    "                overlap = 1\n",
    "            assert overlap > 0, f\"{start} - {end} - {overlap} - {nodes}\"\n",
    "            win_variation += overlap\n",
    "            for row in nodes.itertuples():\n",
    "                if row.sample == \"T2T.\" + backbone:\n",
    "                    continue\n",
    "                elif complement is not None and row.sample.endswith(complement):\n",
    "                    pop_var[\"cmpl_abs\"] += row.node_length\n",
    "                    color_lut[\"cmpl\"] = \"#000000\"\n",
    "                else:\n",
    "                    pop_var[row.super_pop + \"_abs\"] += row.node_length\n",
    "                    color_lut[row.super_pop] = row.color\n",
    "        pop_var[\"win_idx\"] = win_idx\n",
    "        pop_var[\"var_abs\"] = win_variation\n",
    "        pop_variation.append(pop_var)\n",
    "    \n",
    "    plot_df = pd.DataFrame.from_records(pop_variation, index=\"win_idx\")\n",
    "    plot_df.fillna(0, inplace=True)\n",
    "    for column in plot_df.columns:\n",
    "        if not column.endswith(\"abs\"):\n",
    "            continue\n",
    "        new_col = column.replace(\"_abs\", \"_log\")\n",
    "        plot_df[new_col] = (plot_df[column] + 1).astype(float).apply(np.log10)\n",
    "        \n",
    "    fig, ax = plt.subplots(figsize=(10,4))\n",
    "    \n",
    "    #ax.scatter(\n",
    "    #    plot_df.index,\n",
    "    #    plot_df[\"var_log\"],\n",
    "    #    color=\"grey\",\n",
    "    #    label=\"(window variation)\",\n",
    "    #    ls=\"dotted\"\n",
    "    #)\n",
    "    \n",
    "    marker_style = \"x\" if complement == \"X\" else \"v\"\n",
    "    try:\n",
    "        nz_sub = plot_df[\"cmpl_log\"] > 0\n",
    "        ax.scatter(\n",
    "            plot_df.index[nz_sub],\n",
    "            plot_df.loc[nz_sub, \"cmpl_log\"],\n",
    "            color=\"grey\",\n",
    "            label=complement,\n",
    "            marker=marker_style\n",
    "            \n",
    "        )\n",
    "    except KeyError:\n",
    "        assert complement is None\n",
    "        print(\"No homolog\")\n",
    "    \n",
    "    for spop in color_lut.keys():\n",
    "        if spop == \"cmpl\":\n",
    "            continue\n",
    "        try:\n",
    "            nz_sub = plot_df[f\"{spop}_log\"] > 0\n",
    "            ax.scatter(\n",
    "                plot_df.index[nz_sub],\n",
    "                plot_df.loc[nz_sub, f\"{spop}_log\"],\n",
    "                color=color_lut[spop],\n",
    "                label=spop,\n",
    "                ls=\"solid\"\n",
    "            )\n",
    "        except KeyError:\n",
    "            continue\n",
    "            \n",
    "    ax.spines[\"top\"].set_visible(False)\n",
    "    ax.spines[\"right\"].set_visible(False)\n",
    "    \n",
    "    ax.set_xticks(x_ticks)\n",
    "    ax.set_xticklabels(x_ticklabels, fontsize=8)\n",
    "    \n",
    "    ax.set_yticks([1, 2, 3, 4, 5])\n",
    "    ax.set_yticklabels([\"0.01\", \"0.1\", \"1\", \"10\", \"100\"])\n",
    "    \n",
    "    ax.set_xlabel(\n",
    "        (\n",
    "            f\"Genomic location on reference (Mbp)\\n\"\n",
    "            f\"(window size {BIN_LABEL} kbp, shift by {SHIFT_LABEL} kbp)\"\n",
    "        ),\n",
    "        fontsize=12\n",
    "    )\n",
    "    ax.set_ylabel(\"Total ALT allele length\\n(log10-scaled, kbp)\", fontsize=12)\n",
    "    \n",
    "    if complement is None:\n",
    "        ax.set_title(\n",
    "            f'Graph reference (\"backbone\"): {backbone}, plus {num_samples} assemblies'\n",
    "            f'\\n(Total allele length threshold > {size_t} kbp)'\n",
    "        )\n",
    "    else:\n",
    "        ax.set_title(\n",
    "            f'Graph reference (\"backbone\"): {backbone}, plus {complement} and 2*{num_samples} assemblies'\n",
    "            f'\\n(Total allele length threshold > {size_t} kbp)'\n",
    "        )\n",
    "    \n",
    "    ax.legend()\n",
    "            \n",
    "    if save_plots:\n",
    "        out_base_name = f\"FigSX_par1-var_ref-{backbone}_{num_samples}assm\"\n",
    "        plt.savefig(\n",
    "            out_dir / pl.Path(out_base_name + '.pdf'),\n",
    "            bbox_inches='tight',\n",
    "            facecolor=\"w\", transparent=None,\n",
    "            #bbox_extra_artists=[cmap_legend]\n",
    "        )\n",
    "        plt.savefig(\n",
    "            out_dir / pl.Path(out_base_name + '.png'),\n",
    "            bbox_inches='tight',\n",
    "            facecolor=\"w\", transparent=None,\n",
    "            #bbox_extra_artists=[cmap_legend],\n",
    "            dpi=150\n",
    "        )\n",
    "    plt.close()\n",
    "\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
